From 2720bd6759320a5aecbf08a1925dbee2eb87c0cf Mon Sep 17 00:00:00 2001
From: Piotr Nowak <pn@semihalf.com>
Date: Thu, 27 Sep 2012 09:55:22 +0200
Subject: [PATCH 312/609] PM, sflash: Add support to SFlash resume on
 ArmadaXP-B0.

Call SFlash reinitialization routine on board wake-up scope.

Also:
    - Add support to soldered SPI Flash 25Q128 on DB-78460HE-BP
    board.

Signed-off-by: Piotr Nowak <pn@semihalf.com>
Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 .../arm/plat-armada/mv_drivers_lsp/mv_mtd/sflash.c |   17 +++++++++++++-
 arch/arm/plat-armada/mv_hal/sflash/mvSFlash.c      |   24 ++++++++++++++++++++
 arch/arm/plat-armada/mv_hal/sflash/mvSFlashSpec.h  |    5 ++++
 3 files changed, 45 insertions(+), 1 deletion(-)

diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_mtd/sflash.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_mtd/sflash.c
index b4da128..3458e99 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_mtd/sflash.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_mtd/sflash.c
@@ -381,7 +381,22 @@ static int sflash_suspend(struct mtd_info *mtd)
 
 static void sflash_resume(struct mtd_info *mtd)
 {
-	DB(printk("\nINFO: %s called - DUMMY", __FUNCTION__));
+	struct map_info *map = mtd->priv;
+	MV_SFLASH_INFO *sflash = map->fldrv_priv;
+	MV_ULONG flags = 0, sflash_in_irq = 0;
+
+	sflash_disable_irqs(flags, sflash_in_irq);
+	if (mvSFlashInit(sflash) != MV_OK)
+	{
+		sflash_enable_irqs(flags, sflash_in_irq);
+		printk(KERN_NOTICE "ERROR: %s - Failed to initialize the SFlash.", __FUNCTION__);
+		kfree(mtd);
+		kfree(sflash);
+		return;
+	}
+	sflash_enable_irqs(flags, sflash_in_irq);
+
+	printk(KERN_NOTICE "Resuming serial Flash succeeded\n");
 }
 
 static int sflash_block_isbad (struct mtd_info *mtd, loff_t ofs)
diff --git a/arch/arm/plat-armada/mv_hal/sflash/mvSFlash.c b/arch/arm/plat-armada/mv_hal/sflash/mvSFlash.c
index 1ec3a86..3c97d7c 100644
--- a/arch/arm/plat-armada/mv_hal/sflash/mvSFlash.c
+++ b/arch/arm/plat-armada/mv_hal/sflash/mvSFlash.c
@@ -176,6 +176,30 @@ static MV_SFLASH_DEVICE_PARAMS sflash[] = {
      MV_M25P128_MAX_FAST_SPI_FREQ,
      MV_M25P128_FAST_READ_DUMMY_BYTES
     },
+    /* ST 25Q128 SPI flash, 16MB, 64 sectors of 256K each */
+    {
+     MV_M25P_WREN_CMND_OPCD,
+     MV_M25P_WRDI_CMND_OPCD,
+     MV_M25P_RDID_CMND_OPCD,
+     MV_M25P_RDSR_CMND_OPCD,
+     MV_M25P_WRSR_CMND_OPCD,
+     MV_M25P_READ_CMND_OPCD,
+     MV_M25P_FAST_RD_CMND_OPCD,
+     MV_M25P_PP_CMND_OPCD,
+     MV_M25P_SE_CMND_OPCD,
+     MV_M25P_BE_CMND_OPCD,
+     MV_M25P_RES_CMND_OPCD,
+     MV_SFLASH_NO_SPECIFIC_OPCD,    /* power save not supported */
+     MV_M25P128_SECTOR_SIZE,
+     MV_M25P128_SECTOR_NUMBER,
+     MV_M25P_PAGE_SIZE,
+     "ST 25Q128",
+     MV_M25PXXX_ST_MANF_ID,
+     MV_M25Q128_DEVICE_ID,
+     MV_M25Q128_MAX_SPI_FREQ,
+     MV_M25Q128_MAX_FAST_SPI_FREQ,
+     MV_M25Q128_FAST_READ_DUMMY_BYTES
+    },
     /* Macronix MXIC MX25L6405 SPI flash, 8MB, 128 sectors of 64K each */
     {
      MV_MX25L_WREN_CMND_OPCD,
diff --git a/arch/arm/plat-armada/mv_hal/sflash/mvSFlashSpec.h b/arch/arm/plat-armada/mv_hal/sflash/mvSFlashSpec.h
index 88cc3d1..c9b33b8 100644
--- a/arch/arm/plat-armada/mv_hal/sflash/mvSFlashSpec.h
+++ b/arch/arm/plat-armada/mv_hal/sflash/mvSFlashSpec.h
@@ -122,6 +122,11 @@ extern "C" {
 #define     MV_M25PX64_MAX_FAST_SPI_FREQ         50000000    /* 50MHz */
 #define     MV_M25PX64_FAST_READ_DUMMY_BYTES     1
 
+#define     MV_M25Q128_DEVICE_ID                 0xBA18
+#define     MV_M25Q128_MAX_SPI_FREQ              20000000    /* 20MHz */
+#define     MV_M25Q128_MAX_FAST_SPI_FREQ         50000000    /* 50MHz */
+#define     MV_M25Q128_FAST_READ_DUMMY_BYTES     1
+
 /* Sector Sizes and population per device model*/
 #define     MV_M25P32_SECTOR_SIZE               0x10000 /* 64K */
 #define     MV_M25P64_SECTOR_SIZE               0x10000 /* 64K */
-- 
1.7.9.5

